onnxruntime/core/optimizer/pre_shape_node_elimination.cc

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/common/logging/logging.h"
+#include "core/optimizer/rewrite_rule.h"
+#include "core/optimizer/pre_shape_node_elimination.h"
+#include "core/optimizer/utils.h"
+#include "core/graph/graph.h"
+#include "core/graph/graph_utils.h"
+
+namespace onnxruntime {
+
+Status PreShapeNodeElimination::Apply(Graph& graph, Node& node, RewriteRuleEffect& rule_effect, const logging::Logger&) const {
+  // Remove Output Edges.
+  graph_utils::RemoveNodeOutputEdges(graph, node);
+  const Node* p_input_node = graph_utils::GetInputNode(node, 0);
+
+  // Get mutable input node
+  Node& input_node = *graph.GetNode(p_input_node->Index());
+  const int output_idx = graph_utils::GetNodeOutputIndexFromOutputName(input_node, node.MutableInputDefs()[0]->Name());
+
+  // Update InputDefs of the next shape nodes.
+  auto output_nodes = graph.GetMutableConsumerNodes(node.OutputDefs()[0]->Name());
+  for (Node* next_node : output_nodes) {
+    for (auto& input_def : next_node->MutableInputDefs()) {
+      input_def = input_node.MutableOutputDefs()[output_idx];
```

Comment:
wait, we are update every input of the next_node? should we only replace the input that use Cast's output? 

Again shall we add UT case for it also.

Reply:
There is another way to write Unit test, you can refer to https://github.com/microsoft/onnxruntime/pull/16408/files#diff-67e218a0e2228a9f849e01157dcc79815f57b379fbc10f16a8a3cd577e3fb412 as an example.

Reply:
the next node is Shape, and it only takes one input, correct me if im wrong.

Reply:
OK, then let's just update the 1st input:

mutable_consumer.MutableInputDefs()[0] = input_node.MutableOutputDefs()[output_idx];